// @microsoft/kiota-serialization-text@1.0.0-preview.37/dist/es/src/textParseNode.js
import { DateOnly, Duration, parseGuidString, TimeOnly, toFirstCharacterUpper, } from "@microsoft/kiota-abstractions";
export class TextParseNode {
    /**
     *
     */
    constructor(text) {
        this.text = text;
        this.getStringValue = () => this.text;
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        this.getChildNode = (identifier) => {
            throw new Error(TextParseNode.noStructuredDataMessage);
        };
        this.getBooleanValue = () => {
            var _a;
            const value = (_a = this.getStringValue()) === null || _a === void 0 ? void 0 : _a.toLowerCase();
            if (value === "true" || value === "1") {
                return true;
            }
            else if (value === "false" || value === "0") {
                return false;
            }
            return undefined;
        };
        this.getNumberValue = () => Number(this.text);
        this.getGuidValue = () => parseGuidString(this.text);
        this.getDateValue = () => new Date(Date.parse(this.text));
        this.getDateOnlyValue = () => DateOnly.parse(this.getStringValue());
        this.getTimeOnlyValue = () => TimeOnly.parse(this.getStringValue());
        this.getDurationValue = () => Duration.parse(this.getStringValue());
        this.getCollectionOfPrimitiveValues = () => {
            throw new Error(TextParseNode.noStructuredDataMessage);
        };
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        this.getCollectionOfEnumValues = (type) => {
            throw new Error(TextParseNode.noStructuredDataMessage);
        };
        this.getEnumValue = (type) => {
            return type[toFirstCharacterUpper(this.text)];
        };
        if (this.text &&
            this.text.length > 1 &&
            this.text.charAt(0) === '"' &&
            this.text.charAt(this.text.length - 1) === '"') {
            this.text = this.text.substring(1, this.text.length - 2);
        }
    }
    getByteArrayValue() {
        const strValue = this.getStringValue();
        if (strValue && strValue.length > 0) {
            return Buffer.from(strValue, "base64").buffer;
        }
        return undefined;
    }
    /* eslint-disable @typescript-eslint/no-unused-vars */
    getCollectionOfObjectValues(parsableFactory) {
        throw new Error(TextParseNode.noStructuredDataMessage);
    }
    /* eslint-disable @typescript-eslint/no-unused-vars */
    getObjectValue(parsableFactory) {
        throw new Error(TextParseNode.noStructuredDataMessage);
    }
}
TextParseNode.noStructuredDataMessage = "text does not support structured data";
