// @microsoft/kiota-http-fetchlibrary@1.0.0-preview.39/dist/es/src/middlewares/options/headersInspectionOptions.js
import { Headers } from "@microsoft/kiota-abstractions";
export const HeadersInspectionOptionsKey = "HeadersInspectionOptionsKey";
/**
 * @class
 * @implements RequestOption
 * Options
 * Options to inspect headers
 */
export class HeadersInspectionOptions {
    /**
     * @public
     * @getter
     * Gets the request headers
     * @returns the request headers
     */
    getRequestHeaders() {
        return this.requestHeaders;
    }
    /**
     * @public
     * @getter
     * Gets the response headers
     * @returns the response headers
     */
    getResponseHeaders() {
        return this.responseHeaders;
    }
    /**
     * @public
     * @constructor
     * To create an instance of HeadersInspectionOptions
     * @param {HeadersInspectionOptionsParams} [options = {}] - The headers inspection options value
     * @returns An instance of HeadersInspectionOptions
     * @example const options = new HeadersInspectionOptions({ inspectRequestHeaders: true, inspectResponseHeaders: true });
     */
    constructor(options = {}) {
        var _a, _b;
        this.requestHeaders = new Headers();
        this.responseHeaders = new Headers();
        this.inspectRequestHeaders = (_a = options.inspectRequestHeaders) !== null && _a !== void 0 ? _a : false;
        this.inspectResponseHeaders = (_b = options.inspectResponseHeaders) !== null && _b !== void 0 ? _b : false;
    }
    getKey() {
        return HeadersInspectionOptionsKey;
    }
}
