// @microsoft/kiota-http-fetchlibrary@1.0.0-preview.39/dist/es/src/middlewares/options/userAgentHandlerOptions.js
/**
 * -------------------------------------------------------------------------------------------
 * Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.
 * See License in the project root for license information.
 * -------------------------------------------------------------------------------------------
 */
import { libraryVersion } from "./version";
export const UserAgentHandlerOptionsKey = "UserAgentHandlerOptionKey";
/**
 * Represents the options for the UserAgentHandler.
 */
export class UserAgentHandlerOptions {
    getKey() {
        return UserAgentHandlerOptionsKey;
    }
    /**
     * @public
     * @constructor
     * To create an instance of UserAgentHandlerOptions
     * @param {UserAgentHandlerOptionsParams} [options = {}] - The options for the UserAgentHandler
     * @example	const options = new UserAgentHandlerOptions({ enable: false });
     */
    constructor(options = {}) {
        var _a, _b, _c;
        this.enable = (_a = options.enable) !== null && _a !== void 0 ? _a : true;
        this.productName = (_b = options.productName) !== null && _b !== void 0 ? _b : "kiota-typescript";
        this.productVersion = (_c = options.productVersion) !== null && _c !== void 0 ? _c : libraryVersion;
    }
}
